libraries/standard/http/utest/http_send_utest.c

```diff
@@ -0,0 +1,1507 @@
+
+/* The test sets this variable to indicate at which call count of transport send
+ * to return an error from. */
+static uint8_t sendErrorCall = 0;
+
+/* The network data to receive. */
+static uint8_t * pNetworkData = NULL;
+/* The length of the network data to receive. */
+static size_t networkDataLen = 0;
+
+/* The number of bytes to send in the first call to the transport receive
+ * interface. */
+static size_t firstPartBytes = 0;
+/* The count of times a test invoked the transport receive interface. */
+static uint8_t recvCurrentCall = 0;
+
+/* The test sets this variable to indcate which call count count of transport
+ * receive to return an error from. */
+static uint8_t recvStopCall = 0;
+/* The count of times a mocked http_parser_execute callback has been invoked. */
+static uint8_t httpParserExecuteCallCount;
+
+/* The error to set to the parsing context when the http_parser_execute_error
+ * callback is invoked. */
+static enum http_errno httpParsingErrno;
+
+/* Response shared among the tests. */
+static HTTPResponse_t response = { 0 };
+/* Transport interface shared among the tests. */
+static HTTPTransportInterface_t transportInterface = { 0 };
+/* Request headers shared among the tests. */
+static HTTPRequestHeaders_t requestHeaders = { 0 };
+/* Header parsing callback shared among the tests. */
+static HTTPClient_ResponseHeaderParsingCallback_t headerParsingCallback = { 0 };
+
+/* Application callback for intercepting the headers during the parse of a new
+ * response from the mocked network interface. */
+static void onHeaderCallback( void * pContext,
+                              const uint8_t * fieldLoc,
+                              size_t fieldLen,
+                              const uint8_t * valueLoc,
+                              size_t valueLen,
+                              uint16_t statusCode )
+{
+    ( void ) pContext;
+    ( void ) statusCode;
+
+    if( strncmp( ( const char * ) fieldLoc, "Connection", fieldLen ) == 0 )
+    {
+        if( strncmp( ( const char * ) valueLoc, "keep-alive", valueLen ) == 0 )
+        {
+            hasConnectionKeepAlive = 1;
+        }
+        else if( strncmp( ( const char * ) valueLoc, "close", valueLen ) == 0 )
+        {
+            hasConnectionClose = 1;
+        }
+    }
+    else if( strncmp( ( const char * ) fieldLoc, "Content-Length", fieldLen ) == 0 )
+    {
+        contentLength = strtoul( ( const char * ) valueLoc, NULL, 10 );
+    }
+
+    headerCallbackCount++;
+}
+
+/* Successful application transport send interface. */
+static int32_t transportSendSuccess( HTTPNetworkContext_t * pContext,
+                                     const void * pBuffer,
+                                     size_t bytesToWrite )
+{
+    ( void ) pContext;
+
+    if( checkContentLength == 1U )
+    {
+        if( sendCurrentCall == 0U )
+        {
+            size_t contentLengthAndHeaderEndLen = HTTP_TEST_REQUEST_PUT_CONTENT_LENGTH_EXPECTED_LENGTH;

```

Comment:
is this only done for PUT requests?

Reply:
Yes because that is the only request I have in the tests that send a body. 
It made sense to check only once the correct content-length because further tests would be redundant..